008-dplyr-retain-and-exclude.md

```diff
@@ -0,0 +1,855 @@
+
+- `filter()` is optimized for the case of *retaining* rows, but you are
+  just as likely to try and use it for *excluding* rows. Using
+  `filter()` to exclude rows forces you to confront complex boolean
+  logic and explicitly handle missing values, which is difficult to
+  teach, error prone to write, and hard to understand when you come back
+  to it in the future.
+
+- `filter()` combines `,` separated conditions with `&` because this
+  covers the majority of the cases. But if you’d like to combine
+  conditions with `|`, then you have to introduce parentheses around
+  your conditions and combine them into one large condition separated by
+  `|`, reducing readability.
+
+## Solution
+
+To address these issues, we propose two new families of dplyr verbs:
+
+``` r
+# Data frame functions
+retain(.data, ..., .by = NULL, .missing = FALSE)
+exclude(.data, ..., .by = NULL, .missing = FALSE)
+
+# Vector functions
+when_any(..., missing = NULL)
+when_all(..., missing = NULL)
+```
+
+For `retain()` and `exclude()`:
+
+- Both combine conditions with `&`.
+
+- Both treat `NA` as `FALSE`.
+
+  - As we will see, having `exclude()` work in this way simplifies many
+    cases of using `filter()` to exclude rows.
+
+  - `.missing = TRUE` opts in to treating `NA` like `TRUE`. For
+    `retain()`, this retains missing values. For `exclude()`, this
+    excludes missing values.
+
+For `when_any()` and `when_all()`:
+
+- These are equivalents to `pmin()` and `pmax()`, but applied to `any()`
+  and `all()`. They could be called `pany()` and `pall()`, but these
+  names are friendlier.
+
+- `when_any()` combines conditions with `|`. As we will see, this is
+  particularly useful in combination with `retain()`.
+
+- `when_all()` combines conditions with `&`.
+
+- `missing = NULL` propagates `NA` through according to the typical `&`
+  and `|` rules. Propagating missing values by default combines well
+  `retain()` and `exclude()`.
+
+- These functions can be used anywhere, not just in `retain()` and
+  `exclude()`.
+
+- They do have the potential to be confused with `if_any()` and
+  `if_all()`, which apply a function to a selection of columns but
+  otherwise operate similarly.
+
+Notably, `filter()` would roughly alias to `retain()` and would not be
+superseded or deprecated in any way. We recognize that there is too much
+existing code out there for this to be possible. That said, all dplyr
+documentation would be updated to use `retain()` or `exclude()`.
+
+## Implementation
+
+Try for yourself at:
+
+``` r
+pak::pak("tidyverse/dplyr@feature/retain-exclude")
+```
+
+Note that this implementation is not finished yet, and may have bugs.
+
+## Motivation
+
+### `filter()` ambiguity
+
+The word “filter” has two meanings depending on the context:
+
+- Filter *for* some values you want to keep
+
+- Filter *out* some values you want to drop
+
+This is unfortunate and can make `filter()` hard to teach. We believe
```

Comment:
I expected the connection to SQL's `FILTER` to come up around here-ish, i.e. it's not like `dplyr::filter()` exists in a vacuum. But it's true that a lot of dplyr users aren't approaching it through a SQL lens. And I do see that this point is raised at the very end.

Reply:
I think I like the current placement of the SQL discussion so that we can keep the focus on the R side of the API, but I did change the header down at the end to be `### SQL and dbplyr` so readers can more easily see that we have a section about SQL.